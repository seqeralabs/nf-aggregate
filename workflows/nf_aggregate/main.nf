//
// WORKFLOW: Run main seqeralabs/nf-aggregate workflow
//

include { SEQERA_RUNS_DUMP     } from '../../modules/local/seqera_runs_dump'
include { PLOT_RUN_GANTT       } from '../../modules/local/plot_run_gantt'
include { MULTIQC              } from '../../modules/nf-core/multiqc'
include { paramsSummaryMultiqc } from '../../subworkflows/local/utils_nf_aggregate'
include { getProcessVersions   } from '../../subworkflows/local/utils_nf_aggregate'
include { getWorkflowVersions  } from '../../subworkflows/local/utils_nf_aggregate'
include { paramsSummaryMap     } from 'plugin/nf-validation'

workflow NF_AGGREGATE {

    take:
    ids                   // channel: run ids read in from --input
    multiqc_custom_config // channel: user specified custom config file used by MultiQC
    multiqc_logo          // channel: logo rendered in MultiQC report
    seqera_api_endpoint   //     val: Seqera Platform API endpoint URL
    skip_run_gantt        //     val: Skip GANTT chart creation for each run
    skip_multiqc          //     val: Skip MultiQC

    main:

    ch_versions = Channel.empty()
    ch_multiqc_files = Channel.empty()

    //
    // MODULE: Fetch run information via the Seqera CLI
    //
    SEQERA_RUNS_DUMP (
        ids,
        seqera_api_endpoint
    )
    ch_versions = ch_versions.mix(SEQERA_RUNS_DUMP.out.versions.first())

    //
    // MODULE: Generate Gantt chart for workflow execution
    //
    SEQERA_RUNS_DUMP
        .out
        .run_dump
        .filter { 
            meta, run_dir ->
                meta.fusion && !params.skip_run_gantt 
        }
        .set { ch_runs_for_gantt }

    PLOT_RUN_GANTT (
        ch_runs_for_gantt
    )
    ch_versions = ch_versions.mix(PLOT_RUN_GANTT.out.versions.first())

    //
    // Collate software versions
    //
    ch_versions
        .unique()
        .map { getProcessVersions(it) }
        .mix(Channel.of(getWorkflowVersions()))
        .collectFile(storeDir: "${params.outdir}/pipeline_info", name: 'collated_software_mqc_versions.yml', newLine: true)
        .set { ch_collated_versions }

    //
    // MODULE: MultiQC
    //
    ch_multiqc_report = Channel.empty()
    if (!skip_multiqc) {
        ch_multiqc_config = Channel.fromPath("$projectDir/workflows/nf_aggregate/assets/multiqc_config.yml", checkIfExists: true)
        summary_params = paramsSummaryMap(workflow, parameters_schema: "nextflow_schema.json")
        ch_workflow_summary = Channel.value(paramsSummaryMultiqc(summary_params))
        ch_multiqc_files = ch_multiqc_files.mix(ch_workflow_summary.collectFile(name: 'workflow_summary_mqc.yaml'))
        ch_multiqc_files = ch_multiqc_files.mix(ch_collated_versions)
        ch_multiqc_files = ch_multiqc_files.mix(SEQERA_RUNS_DUMP.out.run_dump.collect{it[1]})
        MULTIQC (
            ch_multiqc_files.collect(),
            ch_multiqc_config.toList(),
            multiqc_custom_config.toList(),
            multiqc_logo.toList()
        )
        ch_multiqc_report = MULTIQC.out.report
    }

    emit:
    multiqc_report = ch_multiqc_report
    versions       = ch_versions
}
