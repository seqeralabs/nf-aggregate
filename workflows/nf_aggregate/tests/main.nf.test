nextflow_workflow {

    name "Test Workflow NF_AGGREGATE"
    script "../main.nf"
    workflow "NF_AGGREGATE"

    tag "workflows"
    tag "workflows_nf_aggregate"

    stage {
        symlink "workflows/nf_aggregate/assets/multiqc_config.yml"
    }

    test("Should run without failures") {

        when {
            workflow {
                """
                input[0] = ['id': '4Bi5xBK6E2Nbhj', 'workspace': 'community/showcase']
                input[1] = []
                input[2] = []
                input[3] = 'https://api.tower.nf'
                input[4] = false
                input[5] = false
                """
            }
        }

        then {
            assertAll(
                { assert snapshot(
                    workflow.out.versions
                ).match()},
                { assert path(workflow.out.multiqc_report.get(0)).getText().contains("nf-core/rnaseq_4Bi5xBK6E2Nbhj") },
                { assert workflow.success }
            )
        }

    }

}
