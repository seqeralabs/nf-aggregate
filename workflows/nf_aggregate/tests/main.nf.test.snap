{
    "Should run without failures": {
        "content": [
            [
                "versions.yml:md5,8ab950fb3e4c32ce4745551fa9a4b8e7",
                "versions.yml:md5,f3cf523314479fcfc842e81116af8ab4"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-02T13:45:38.806631609"
    }
}